include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -1203,68 +1204,82 @@ irr_glsl_LightSample irr_bsdf_cos_generate(in MC_precomputed_t precomp, in instr
+			{
+				//TODO G1, G2_over_G1, ndf_val
+			} CASE_END
 #endif //NDF_GGX
 
 #ifdef NDF_BECKMANN
-		CASE_BEGIN(ndf, NDF_BECKMANN)
-		{
-			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
-			G1 = irr_glsl_smith_G1(lambdaV);
-			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL * s.TdotL, s.BdotL * s.BdotL, s.NdotL * s.NdotL, ax2, ay2);
-
-			const float TdotH2 = out_microfacet.TdotH * out_microfacet.TdotH;
-			const float BdotH2 = out_microfacet.BdotH * out_microfacet.BdotH;
-			const float NdotH2 = out_microfacet.isotropic.NdotH2;
-			ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
-		} CASE_END
+			CASE_BEGIN(ndf, NDF_BECKMANN)
+			{
+				float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+				G1 = irr_glsl_smith_G1(lambdaV);
+				G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL * s.TdotL, s.BdotL * s.BdotL, s.NdotL * s.NdotL, ax2, ay2);
+
+				const float TdotH2 = out_microfacet.TdotH * out_microfacet.TdotH;
+				const float BdotH2 = out_microfacet.BdotH * out_microfacet.BdotH;
+				const float NdotH2 = out_microfacet.isotropic.NdotH2;
+				ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+			} CASE_END
 #endif //NDF_BECKMANN
 
 #ifdef NDF_PHONG
-		CASE_BEGIN(ndf, NDF_PHONG)
-		{
-			//TODO G1, G2_over_G1, ndf_val
-		} CASE_END
+			CASE_BEGIN(ndf, NDF_PHONG)
+			{
+				//TODO G1, G2_over_G1, ndf_val
+			} CASE_END
 #endif //NDF_PHONG
-		CASE_OTHERWISE
-		{}
-		END_CASES
+			CASE_OTHERWISE
+			{}
+			END_CASES
+		}
 	} else
 #endif
 	{} //empty braces for `else`
 
-	rem *= G2_over_G1;
-	pdf *= (G1 * ndf_val);
-	out_remainder = weight*rem;
-	out_pdf = weight*pdf; // TODO not multiplu pdf by weight??
+	if (out_valid_sample)
+	{
+		rem *= G2_over_G1;
+		pdf *= (G1 * ndf_val);
+		out_remainder = weight * rem;
+		out_pdf = weight * pdf; // TODO not multiplu pdf by weight??
+	}
 
 	return s;
 }
 
 vec3 runGenerateAndRemainderStream(in MC_precomputed_t precomp, in instr_stream_t gcs, in instr_stream_t rnps, in vec3 rand, out float out_pdf, out irr_glsl_LightSample out_smpl)
 {
+	out_pdf = 0.0;
+
 	instr_t generator;
 	vec3 generator_rem;
 	float generator_pdf;
 	irr_glsl_AnisotropicMicrofacetCache microfacet;
 	uint generator_rnpOffset;
-	irr_glsl_LightSample s = irr_bsdf_cos_generate(precomp, gcs, rand, generator_rem, generator_pdf, microfacet, generator_rnpOffset);
-	eval_and_pdf_t eval_pdf = irr_bsdf_eval_and_pdf(precomp, rnps, s, generator_rnpOffset, microfacet);
-	bxdf_eval_t acc = eval_pdf.rgb;
-	float restPdf = eval_pdf.a;
-	float pdf = generator_pdf + restPdf;
-
+	bool valid_sample;
+	irr_glsl_LightSample s = irr_bsdf_cos_generate(precomp, gcs, rand, generator_rem, generator_pdf, microfacet, generator_rnpOffset, valid_sample);
 	out_smpl = s;
-	out_pdf = pdf;
 
-	vec3 rem = generator_rem/(1.0 + restPdf/generator_pdf) + acc/pdf;
+	vec3 rem = vec3(0.0);
+	if (valid_sample)
```

Comment:
nope... you're supposed to check if samples are valid inside `irr_bsdf_cos_generate` for each BSDF, separately